### PR TITLE
git_backend: Support shallow git repositories

### PR DESCRIPTION
**Pull request by Veykril**
Fixes https://github.com/martinvonz/jj/issues/675

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

**Comment by Veykril** [comment]
Have not tested it with partial repos yet (also only did some light manual testing). Unsure how to write a proper test for this, so if anyone has any ideas I am all ears :)

File: cli/src/commands/git/clone.rs [hunk 1]
```diff
@@ -57,6 +57,9 @@ pub struct GitCloneArgs {
     /// Whether or not to colocate the Jujutsu repo with the git repo
     #[arg(long)]
     colocate: bool,
+    /// Create a shallow clone of the given depth.
+    #[arg(long)]
+    depth: Option<usize>,
```
Comment (PhilipMetzger) [inline cli/src/commands/git/clone.rs:62]: Nit: We don't end help message sentences with a period. And this could be 
```suggestion
   #[arg(long, default = 0)]
   depth: usize,
```

File: lib/src/git.rs
Comment (PhilipMetzger) [inline lib/src/git.rs:1]: Nit: I think this belongs to a separate `lib:`/`git:` commit or it could be squashed into the parent. This is for all touched files in `lib/`.

File: lib/src/git.rs [hunk 2]
```diff
@@ -1240,6 +1240,7 @@ pub struct GitFetchStats {
 pub fn fetch(
     mut_repo: &mut MutableRepo,
     git_repo: &git2::Repository,
+    depth: usize,
```
Comment (PhilipMetzger) [inline lib/src/git.rs:1243]: Nit: Isn't a `Option<usize>` more  convienent here?

File: lib/src/git.rs [hunk 3]
```diff
@@ -1260,6 +1261,7 @@ pub fn fetch(
     fetch_options.proxy_options(proxy_options);
     let callbacks = callbacks.into_git();
     fetch_options.remote_callbacks(callbacks);
+    fetch_options.depth(depth.try_into().unwrap_or(0));
     // At this point, we are only updating Git's remote tracking branches, not the
```
Comment (PhilipMetzger) [inline lib/src/git.rs:1265]: Then this could be wrapped into a if let block.

File: lib/src/git_backend.rs
Comment (PhilipMetzger) [inline lib/src/git_backend.rs:1]: Optional: This commit could have the topic `git_backend:` if you want to be consistent with the files history.

File: docs/git-compatibility.md [hunk 4]
```diff
@@ -57,8 +57,7 @@ a comparison with Git, including how workflows are different, see the
   not be lost either.
 * **Partial clones: No.** We use the [libgit2](https://libgit2.org/) library,
   which [doesn't have support for partial clones](https://github.com/libgit2/libgit2/issues/5564).
-* **Shallow clones: No.** We use the [libgit2](https://libgit2.org/) library,
-  which [doesn't have support for shallow clones](https://github.com/libgit2/libgit2/issues/3058).
+* **Shallow clones: Yes.** Shallow commits all have the virtual root commit as their parent.
```
Comment (PhilipMetzger) [inline docs/git-compatibility.md:60]: Nit: This still needs to wrap at 80.

File: docs/git-compatibility.md
Comment (PhilipMetzger) [inline docs/git-compatibility.md:1]: Nit: Also contains some unrelated changes again.

**Review by PhilipMetzger — CHANGES_REQUESTED** [review]
Minor nits and some points for further discussion. I also think the initial PR description was good, but we don't really have guidelines here.

File: lib/src/git_backend.rs [hunk 5]
```diff
@@ -537,10 +538,16 @@ fn commit_from_git_without_root_parent(
...
     );
-    let parents = commit
-        .parents()
-        .map(|oid| CommitId::from_bytes(oid.as_bytes()))
-        .collect_vec();
+    // shallow commits don't have parents their parents actually fetched, so we
+    // discard them here
+    let parents = if is_shallow {
+        vec![]
+    } else {
```
Comment (yuja) [inline lib/src/git_backend.rs:545]: Maybe this means that the shallow-boundary commit was rewritten, its parents would be permanently lost?

That might be okay for experimental support of shallow clones, but we'll need to find a better workaround later.

File: docs/git-compatibility.md
Comment (Veykril) [reply docs/git-compatibility.md:1]: Dammit you have a lot of trailing whitespace in your repo 😅

File: lib/src/git_backend.rs (hunk 5 above)
Comment (Veykril) [reply lib/src/git_backend.rs:545]: As in, if someone tries to unshallow their repository we won't pick up those changes? I'll have to check but I imagine we can work around that by marking shallow commits or so.

Either way I'd also say for initial support that would be be probably fine.

File: docs/git-compatibility.md
Comment (martinvonz) [reply docs/git-compatibility.md:1]: IMO, cleanups are welcome, btw, but they should ideally be separate commits.

The better solution is to set up some markdown formatter that runs as a GitHub Action.

File: lib/src/git_backend.rs (hunk 5 above)
Comment (martinvonz) [reply lib/src/git_backend.rs:545]: Yes, not crashing is at least better than crashing.

If the repo is later unshallowed (or just deepened a bit), we would somehow want to detect that so we can rebuild our commit index. That seems to suggest some kind of weird interaction between the commit backend and the index backend that doesn't currently exist (the index backend is agnostic to the kind of commit backend, it doesn't treat the Git backend specially in any way).

Comment (yuja) [reply lib/src/git_backend.rs:545]: > if someone tries to unshallow their repository we won't pick up those changes?

We might get panic somewhere because parents aren't indexed.

> Either way I'd also say for initial support that would be be probably fine.

+1. TODO comments will also help.

Comment (Veykril) [reply lib/src/git_backend.rs:545]: Okay I just checked, we do in fact panic if the repo is deepened (`lib\src\default_index\composite.rs:472:49` in my testing just now)

**Comment by Veykril** [comment]
Okay this is ready for proper review now. I'm leaving deepening/unshallowing for a separate issue (will open an issue for that after this is merged)

File: lib/src/git_backend.rs [hunk 6]
```diff
@@ -859,6 +867,8 @@ fn import_extra_metadata_entries_from_heads(
     head_ids: &HashSet<&CommitId>,
     uses_tree_conflict_format: bool,
 ) -> BackendResult<()> {
+    let shallow_commits = git_repo.shallow_commits().ok().flatten();
```
Comment (yuja) [inline lib/src/git_backend.rs:870]: nit: better to map error to `BackendError::Other`, but that can be addressed later. We'll get another read error anyway.

File: lib/tests/test_git.rs [hunk 7]
```diff
@@ -3468,3 +3468,99 @@ ignoreThisSection = foo
...
+    let e = empty_git_commit(&git_repo, "refs/heads/main", &[&c]);
+
+    git_repo.set_head("refs/heads/main").unwrap();
+
+    let make_shallow = |repo, mut shallow_commits: Vec<_>| {
+        let shallow_file = get_git_backend(repo).git_repo().shallow_file();
+        shallow_commits.sort();
+        let mut buf = Vec::<u8>::new();
+        for commit in shallow_commits {
+            std::io::Write::write_fmt(&mut buf, format_args!("{commit}\n")).unwrap();
```
Comment (yuja) [inline lib/tests/test_git.rs:3513]: nit: `use std::io::Write` and `writeln!()`.

File: lib/tests/test_git.rs [hunk 8]
```diff
@@ -3468,3 +3468,99 @@ ignoreThisSection = foo
...
+    };
+    assert_eq!(*repo.view().heads(), expected_heads);
+
+    let parents = |store: &Arc<jj_lib::store::Store>, commit| {
+        let commit = store.get_commit(&jj_id(commit)).unwrap();
+        commit
+            .parents()
+            .map(Result::unwrap)
+            .map(|c| c.id().clone())
+            .collect_vec()
```
Comment (yuja) [inline lib/tests/test_git.rs:3537]: nit: `commit.parent_ids().to_vec()`

File: lib/src/git.rs [hunk 9]
```diff
@@ -1240,6 +1241,7 @@ pub struct GitFetchStats {
 pub fn fetch(
     mut_repo: &mut MutableRepo,
     git_repo: &git2::Repository,
+    depth: Option<NonZeroU32>,
```
Comment (yuja) [inline lib/src/git.rs:1244]: nit: I would insert `depth` argument somewhere around `callbacks` and `git_settings`. It's not a mandatory option, and we might want to extract a parameter struct later.

File: lib/src/git.rs [hunk 10]
```diff
@@ -1260,6 +1262,9 @@ pub fn fetch(
     fetch_options.proxy_options(proxy_options);
     let callbacks = callbacks.into_git();
     fetch_options.remote_callbacks(callbacks);
+    if let Some(depth) = depth.and_then(|depth| depth.get().try_into().ok()) {
```
Comment (yuja) [inline lib/src/git.rs:1265]: nit: better to clamp at `i32::MAX` than ignoring overflow

File: cli/src/commands/git/clone.rs [hunk 11]
```diff
@@ -60,6 +61,11 @@ pub struct GitCloneArgs {
     /// Whether or not to colocate the Jujutsu repo with the git repo
     #[arg(long)]
     colocate: bool,
+    /// Create a shallow clone of the given depth
+    // We accept a depth of 0 for parity with git's --depth option.
+    // The option is here to not render a default = 0 in the help message.
```
Comment (yuja) [inline cli/src/commands/git/clone.rs:66]: nit: I don't think we need to copy the exact git CLI behavior. If `--depth 0` doesn't make sense, I would make it an error.

File: cli/src/commands/git/clone.rs
Comment (yuja) [inline cli/src/commands/git/clone.rs:1]: Can you add CLI tests of `jj git clone --depth`?

**Review by yuja — COMMENTED** [review]
LGTM, thanks!

**Review by PhilipMetzger — APPROVED** [review]
LG after addressing Yuya's nits.

File: cli/src/commands/git/clone.rs
Comment (Veykril) [reply cli/src/commands/git/clone.rs:1]: It looks like local clones do not support shallow clones in git2, yielding
> shallow fetch is not supported by the local transport

Comment (yuja) [reply cli/src/commands/git/clone.rs:1]: Ok, it wouldn't be easy to write tests, then. (we could test that non-zero `--depth` is passed to libgit2, though.)

Thanks for checking.

File: lib/src/git.rs
Comment (PhilipMetzger) [inline lib/src/git.rs:1]: Nit: squash into previous.

Comment (Veykril) [reply lib/src/git.rs:1]: Squash which into which, all 3 commits touch this file

Comment (PhilipMetzger) [reply lib/src/git.rs:1]: These changes address Yuya's nits from the second commit, so move them there.
